### DISCUSSION TITLE
Wrong vector size when attaching word2vec to transformer model

### DISCUSSION THREAD

**Discussion by Geckao** [explosion/spaCy#7917] [Help: Coding & Implementations] [answered]
Hello,

I am trying to use the whole pipeline of the latest transformer model (en_core_web_trf) along with the tok2vec component of the en_core_web_lg model, since I need the vectors. I naively tried to attach the tok2vec component at the end of the transformer model, as folows:
`trf = spacy.load('en_core_web_trf')`
`cnn = spacy.load('en_core_web_lg')`
`trf.add_pipe('tok2vec', source=cnn)`
However, I notice that the vector size of the tokens' vectors is 96 instead of the usual 300. What is the issue with this method ?
Thank you for your time !

**Comment by polm** [comment, marked as answer]
`tok2vec` and the static vectors are different things. 

The static vectors are pretrained vectors (from GloVe for the English models) with dimension 300. 

The tok2vec layer is a HashEmbed layer trained on downstream tasks, with the static vectors as an input. Its outputs are dimension 96. 

Normally you can load vectors from another model and just set them on the vocab, but with the Transformer model that will change the way the `Doc.vector` method works and can cause issues. This is how you would do that:

```
nlp = spacy.load("en_core_web_trf")
vecnlp = spacy.load("en_core_web_lg")
nlp.vocab.vectors = vecnlp.vocab.vectors
```

Can you clarify what you're trying to do and why you want both the Transformer and static vectors?

**Reply by Geckao** [reply]
Oh good to know, I thought that the tok2vec vectors were the static ones.
Great that works ! But what kind of issues should I expect ? The Doc.vector of the Transformer is an empty array anyway.

> Can you clarify what you're trying to do and why you want both the Transformer and static vectors?

I need the dependency parsing and NER of the Transformer, as well as static word embeddings. I then compare those embeddings with static ones (generated from words without context) on a database. I am curious though, how can I get the Transformer embeddings from the en_core_web_trf model ?

**Reply by polm** [reply]
Transformer embeddings are stored in `._.trf_data`, along with other info, so you can get the embedding as `doc._.trf_data.tensors`. 

> But what kind of issues should I expect ?

To be honest I'm not familiar enough with the implementation of the Transformers integration to be sure what would happen, but the implementation generally assumes that Transformers models don't have static vectors. It might be totally fine I would be careful about using them this way, just because it might be untested.

If you do run into issues, one thing you could do is register a lookup for the static vectors as a token extension. This would lose integration with native spaCy features like `Doc.vectors` but it would avoid any unwelcome interactions.

**Reply by Geckao** [reply]
> Transformer embeddings are stored in ._.trf_data, along with other info, so you can get the embedding as doc._.trf_data.tensors.

Great, good to know !

> If you do run into issues, one thing you could do is register a lookup for the static vectors as a token extension.

Alright, I guess I will try with the first method to see if it works (since it's so fast to implement) but I will keep in mind that it might cause bugs and that there's a safer way to do it.

Thank you very much for your help !

**Reply by adrianeboyd** [reply]
As a note: I don't think the added vectors will interact with the transformer component at all.
